pixeltable/func/function.py

```diff
@@ -64,37 +73,114 @@ def display_name(self) -> str:
             return self.self_path[len(ptf_prefix):]
         return self.self_path
 
+    @property
+    def is_polymorphic(self) -> bool:
+        return len(self.signatures) > 1
+
+    @property
+    def signature(self) -> Signature:
+        assert not self.is_polymorphic
+        return self.signatures[0]
+
     @property
     def arity(self) -> int:
+        assert not self.is_polymorphic
         return len(self.signature.parameters)
 
+    def _overload_resolution(self, signature_idx: int) -> Function:
+        """
+        Return a new  `Function` that retains just the single signature at index `signature_idx`, and is otherwise
+        identical to this `Function`.
+        """
+        assert 0 <= signature_idx < len(self.signatures)
+
+        # We cache the resolutions in self.__resolutions. This ensures that each resolution is represented globally
+        # by a single Python object. We do this dynamically rather than pre-constructing them in order to avoid
+        # circular complexity in the `Function` initialization logic.
+
+        # Note that it is necessary to create a resolution even if this Function has only one signature, because
+        # there is no guarantee we have seen all the Function's @overload declarations at the time this is called.
+
+        while len(self.__resolutions) <= signature_idx:
+            idx = len(self.__resolutions)
+            resolution = copy(self)
+            resolution.signatures = [self.signatures[idx]]
+            resolution._update_as_overload_resolution(idx)
+            self.__resolutions.append(resolution)
+
+        return self.__resolutions[signature_idx]
+
+    def _update_as_overload_resolution(self, signature_idx: int) -> None:
+        """
+        Subclasses must implement this in order to do any additional work when creating a resolution, beyond
+        simply updating `self.signatures`.
+        """
+        raise NotImplementedError()
+
     def help_str(self) -> str:
-        return self.display_name + str(self.signature)
+        return self.display_name + str(self.signatures[0])
 
     def __call__(self, *args: Any, **kwargs: Any) -> 'pxt.exprs.FunctionCall':
         from pixeltable import exprs
-        bound_args = self.signature.py_signature.bind(*args, **kwargs)
-        self.validate_call(bound_args.arguments)
-        return exprs.FunctionCall(self, bound_args.arguments)
+
+        mono_fn, bound_args = self._bind_to_matching_signature(args, kwargs)
+        return_type = mono_fn.call_return_type(args, kwargs)
+        return exprs.FunctionCall(mono_fn, bound_args, return_type)
+
+    def _bind_to_matching_signature(self, args: Sequence[Any], kwargs: dict[str, Any]) -> tuple[Function, dict[str, Any]]:
```

Comment:
this is basically doing the overload resolution (ie, determine which of the available overloads is a good match).

_overload_resolution() is simply giving you a Function for the resolved signature.

Reply:
See comment below
<!-- Reviewable comment -OEWnioO2Z8JLBgdlCS5:b-32pjdu -->
<!-- Sent from Reviewable.io -->